### PR TITLE
SOLR-16960 Tests should sometimes run with a Tracer (not no-op)

### PR DESCRIPTION
**Pull request by stillalex**
https://issues.apache.org/jira/browse/SOLR-16960

# Description

Took a stab at randomizing the isRecording flag during tests so we get better coverage.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/test-framework/src/java/org/apache/solr/cloud/MiniSolrCloudCluster.java [hunk 1]
```diff
@@ -1299,5 +1301,15 @@ public Builder withTraceIdGenerationDisabled() {
...
       return this;
     }
+
+    /**
+     * It only makes sense to call this if we are using the alwaysOn tracer. this will randomize the
+     * Span::isRecording check for the alwaysOn tracer so we have coverage of all methods that deal
+     * with span creation without having to enable otel module
+     */
+    private void injectRandomRecordingFlag() {
+      boolean isRecording = LuceneTestCase.random().nextBoolean();
```
Comment (dsmiley) [inline solr/test-framework/src/java/org/apache/solr/cloud/MiniSolrCloudCluster.java:1311]: I believe there is a "rarely()" method or something like that.  I think it should be rare that we add this and not merely 50%.

**Review by dsmiley — APPROVED** [review]
+1 thanks.

Progress over perfection, -- this is good, but ideally we would inject the propagating tracer and potentially this recording flag in an independent way that isn't specific to SolrCloud.  Tracing isn't specific to SolrCloud!  In the mean time, it'd be nice if there was a comment to that effect.  People learn systems by reading code; someone might think tracing is SolrCloud specific seeing this (and be wrong).

File: solr/test-framework/src/java/org/apache/solr/cloud/MiniSolrCloudCluster.java (hunk 1 above)
Comment (stillalex) [reply solr/test-framework/src/java/org/apache/solr/cloud/MiniSolrCloudCluster.java:1311]: good suggestion. will move to rarely()

**Comment by stillalex** [comment]
> but ideally we would inject the propagating tracer and potentially this recording flag in an independent way that isn't specific to SolrCloud. Tracing isn't specific to SolrCloud! In the mean time, it'd be nice if there was a comment to that effect. People learn systems by reading code; someone might think tracing is SolrCloud specific seeing this (and be wrong).

we don't disagree. I want to make this available to all testing code, and it's easy enough to do so, the current code serves as an example. what is confusing is not having a clear indication of what is missing. do you want to move this to a better location? please suggest a better place. there is so much testing code around I usually get lost in the options and pick one that looks 'good enough'.

**Comment by stillalex** [comment]
@dsmiley fyi. reworded the javadocs. will merge once the build completes.

**Review by anshumg — APPROVED** [review]

**Pull request merged**